File: backend/api/schemas/submission_schema.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class SubmissionResponse(BaseModel):
    """Response schema for submission data"""
    id: int
    assignment_id: int
    student_id: int
    file_name: str
    file_size: int
    submission_number: int
    submitted_at: str
    is_latest: bool

    # Optional related data
    assignment_name: Optional[str] = None
    student_name: Optional[str] = None

    class Config:
        from_attributes = True
        populate_by_name = True


class SubmitAssignmentResponse(BaseModel):
    """Response schema for assignment submission"""
    success: bool
    message: Optional[str] = None
    submission: Optional[SubmissionResponse] = None


class SubmissionListResponse(BaseModel):
    """Response schema for list of submissions"""
    success: bool
    message: Optional[str] = None
    submissions: List[SubmissionResponse]


class SubmissionHistoryResponse(BaseModel):
    """Response schema for submission history"""
    success: bool
    message: Optional[str] = None
    submissions: List[SubmissionResponse]
    total_submissions: int


class SubmissionDetailResponse(BaseModel):
    """Response schema for single submission details"""
    success: bool
    message: Optional[str] = None
    submission: Optional[SubmissionResponse] = None

File: backend-ts/src/api/schemas/submission.schema.ts
import { z } from 'zod';

/** Submission response schema */
export const SubmissionResponseSchema = z.object({
    id: z.number(),
    assignmentId: z.number(),
    studentId: z.number(),
    fileName: z.string(),
    filePath: z.string(),
    fileSize: z.number(),
    submissionNumber: z.number(),
    submittedAt: z.string(),
    isLatest: z.boolean(),
    studentName: z.string().optional(),
    studentUsername: z.string().optional(),
    assignmentName: z.string().optional(),
});

export type SubmissionResponse = z.infer<typeof SubmissionResponseSchema>;

/** Submit assignment response schema */
export const SubmitAssignmentResponseSchema = z.object({
    success: z.boolean(),
    message: z.string(),
    submission: SubmissionResponseSchema.optional(),
});

export type SubmitAssignmentResponse = z.infer<typeof SubmitAssignmentResponseSchema>;

/** Submission list response schema */
export const SubmissionListResponseSchema = z.object({
    success: z.boolean(),
    message: z.string(),
    submissions: z.array(SubmissionResponseSchema),
});

export type SubmissionListResponse = z.infer<typeof SubmissionListResponseSchema>;

/** Submission history response schema */
export const SubmissionHistoryResponseSchema = z.object({
    success: z.boolean(),
    message: z.string(),
    submissions: z.array(SubmissionResponseSchema),
    totalSubmissions: z.number(),
});

export type SubmissionHistoryResponse = z.infer<typeof SubmissionHistoryResponseSchema>;

/** Submission detail response schema */
export const SubmissionDetailResponseSchema = SubmissionResponseSchema.extend({
    assignmentName: z.string().optional(),
    className: z.string().optional(),
});

export type SubmissionDetailResponse = z.infer<typeof SubmissionDetailResponseSchema>;
